smithy-model/src/main/java/software/amazon/smithy/model/shapes/ServiceShape.java

```diff
@@ -46,11 +89,11 @@ public static Builder builder() {
     @Override
     public Builder toBuilder() {
         return updateBuilder(builder())
-                .version(version)
-                .errors(errors)
-                .rename(rename)
-                .operations(getOperations())
-                .resources(getResources());
+                .version(introducedVersion)
+                .errors(introducedErrors)
+                .rename(introducedRename)
+                .operations(getIntroducedOperations())
```

Comment:
I’m on a phone… sorry… how critical is it that we compute the “introduced” stuff up front? Is t it mostly only used when serializing a model? I’m wondering if we can yank all of this stuff out and then compute it when serializing a model by traversing the model. It could be wrapped in a knowledge index too if we think people might need it for other uses.

Reply:
It's not super critical. It's just a lot harder to do it after the fact.

Reply:
Or maybe these introduced things could be computed lazily. Possibly traits too, but we’d want to benchmark it — to get a trait via getTrait(), you check the current shape then iterate backwards over mixins for example. Each “introduced” method could be computed on demand (e.g., getting version means first check current shape, then check mixins, every time). My idea for moving this to a knowledge index wouldn’t work since all this information is local to shapes. Plus we do want to serialize traits and introduced things even if they repeat one of their mixins as that is an explicitly set value on a shape (something you can’t tell from outside of a shape).

I’m mostly concerned with time and space costs of introducing the additional sets, maps, up front validation, etc. I’m thinking of the use case of loading every AWS model at once and running a selector. The way it’s done now could be the best way and I could be totally wrong though.

Reply:
You actually can do it with a knowledge index, i have updated the pr to do it. It's just super tedious